### ISSUE TITLE
Glitches?

### ISSUE THREAD

**Issue by uXeBoy** [afiskon/fpga-ssd1306-to-vga#1]
OK, successfully compiled the SystemVerilog - the two issues I [mentioned](https://github.com/uXeBoy/VGA1306/issues/1#issuecomment-381301208) still remain though.

**Comment by afiskon** [comment]
Hello @uXeBoy ,

Thanks for your feedback! I doubt something is seriously wrong on the FPGA side since all that configuration does is reading page after page ignoring all commands.

I believe the most likely reasons of the issue are:

- Bad connection. I usually fix all the hardware on the table with insulating tape, just to make sure.
- The display is not quite happy with signal timing. The values I used were taken from http://martin.hinner.info/vga/timing.html and work for my display. Try to change h_pulse, h_bp and other parameters. IIRC you had a bit different values in the original code.
- Particular protocol implementation of the library you are using. E.g. pages sometimes are sent out of order or are filled with white color before redrawing.

Do you have an oscilloscope or logic analyzer to record what exactly is sent to the FPGA? If not you can use Sigrok/PulseView and 8 channel Saleae Logic Clone ( https://sigrok.org/wiki/Noname_Saleae_Logic_clone ). It costs ~7$ on eBay, look for "24mhz 8 channel logic analyzer". Here is an excellent tutorial from PulseView's author https://www.youtube.com/watch?v=dobU-b0_L1I

**Comment by afiskon** [comment]
Also recently I got BlackIce II and ported my code to this board: https://github.com/afiskon/blackice-ii-ssd1306-to-vga/ Maybe it will help to investigate an issue.

It uses internal STM32 MCU to configure the FPGA and generate the video signal. To compile the code STLink v2 programmer and STM32CubeMX software are required.

**Comment by uXeBoy** [comment]
You are right, it is certainly _possible_ that one of those reasons are causing the glitches - will have to investigate.

Could you confirm on your hardware whether drawing a pixel from the MCU at the far right edge of the display (128th column) causes that pixel to be mirrored to the far left edge (1st column)?

**Comment by afiskon** [comment]
Hello @uXeBoy ,

Sorry it took me that long to reply.

> Could you confirm on your hardware whether drawing a pixel from the MCU at the far right edge of the display (128th column) causes that pixel to be mirrored to the far left edge (1st column)?

I started investigating this and then I realized - columns on SSD1306-compatible displays are numbered starting from 0. So yes if you try to draw a pixel in 128th column it will appear in 0th column, on the left.

**Comment by uXeBoy** [comment]
OK, let me rephrase that then... :smiley:

Could you confirm on your hardware whether drawing a pixel from the MCU at the far right edge of the display (column numbered 127) causes that pixel to be mirrored to the far left edge (column numbered 0)?

**Comment by afiskon** [comment]
Frankly I'm having some difficulties to check this. I wrote the following patch for https://github.com/afiskon/blackice-ii-ssd1306-to-vga/ : 

```
diff --git a/stm32/ssd1306/ssd1306_tests.c b/stm32/ssd1306/ssd1306_tests.c
index c2fdbae..2e818cc 100644
--- a/stm32/ssd1306/ssd1306_tests.c
+++ b/stm32/ssd1306/ssd1306_tests.c
@@ -48,9 +48,38 @@ void ssd1306_TestFPS() {
     ssd1306_UpdateScreen();
 }
 
+void ssd1306_TestBorder() {
+    ssd1306_Fill(Black);
+
+    uint32_t start = HAL_GetTick();
+    uint32_t end = start;
+    uint8_t x = 0;
+    uint8_t y = 0;
+    do {
+        ssd1306_DrawPixel(x, y, Black);
+
+        if((y == 0) && (x < 127))
+            x++;
+        else if((x == 127) && (y < 63))
+            y++;
+        else if((y == 63) && (x > 0))
+            x--;
+        else
+            y--;
+
+        ssd1306_DrawPixel(x, y, White);
+        ssd1306_UpdateScreen();
+
+        HAL_Delay(1);
+        end = HAL_GetTick();
+    } while((end - start) < 8000);
+    HAL_Delay(1000);
+}
+
 void ssd1306_TestAll() {
     ssd1306_Init();
     ssd1306_TestFPS();
     HAL_Delay(3000);
+    ssd1306_TestBorder();
     ssd1306_TestFonts();
 }
diff --git a/stm32/ssd1306/ssd1306_tests.h b/stm32/ssd1306/ssd1306_tests.h
index 5bd023d..6d950bd 100644
--- a/stm32/ssd1306/ssd1306_tests.h
+++ b/stm32/ssd1306/ssd1306_tests.h
@@ -1,6 +1,7 @@
 #ifndef __SSD1306_TEST_H__
 #define __SSD1306_TEST_H__
 
+void ssd1306_TestBorder();
 void ssd1306_TestFonts();
 void ssd1306_TestFPS();
 void ssd1306_TestAll();
```

The problem is that when I apply it STLink fails to upload the firmware:

```
st-flash --reset write build/main.bin 0x8000000
st-flash 1.5.0
2018-04-25T20:40:34 INFO common.c: Loading device parameters....
2018-04-25T20:40:34 INFO common.c: Device connected is: L43x/L44x device, id 0x10016435
2018-04-25T20:40:34 INFO common.c: SRAM size: 0xc000 bytes (48 KiB), Flash: 0x40000 bytes (256 KiB) in pages of 2048 bytes
2018-04-25T20:40:34 INFO common.c: Attempting to write 157024 (0x26560) bytes to stm32 address: 134217728 (0x8000000)
Flash page at addr: 0x08026000 erasedEraseFlash - Page:0x10c Size:0x800 
2018-04-25T20:40:36 INFO common.c: Finished erasing 77 pages of 2048 (0x800) bytes
2018-04-25T20:40:36 INFO common.c: Starting Flash write for F2/F4/L4
2018-04-25T20:40:36 INFO flash_loader.c: Successfully loaded flash loader in sram
size: 32768
size: 32768
size: 32768
size: 32768
size: 25952
2018-04-25T20:40:39 INFO common.c: Starting verification of write complete
2018-04-25T20:40:41 ERROR common.c: Verification of flash failed at offset: 131072
stlink_fwrite_flash() == -1
make: *** [Makefile:207: flash] Error 255
```

If I rollback the patch everything is OK once again. I tried to change the wires between STLink and BlackIce II but it didn't help.

I have no idea what's going on :(

**Comment by uXeBoy** [comment]
You could restore iceboot to the BlackIce's STM32, and just interface with an external STM32 instead to work around this complication? or go back to using the iCEstick instead?

**Comment by afiskon** [comment]
OK, I solved this "Verification of flash failed" issue by implementing a simple compression algorithm for FPGA configuration. The current code could be found here: https://github.com/afiskon/blackice-ii-ssd1306-to-vga

Locally I modify it like this:

```
--- a/stm32/ssd1306/ssd1306_tests.c
+++ b/stm32/ssd1306/ssd1306_tests.c
@@ -78,8 +78,8 @@ void ssd1306_TestBorder() {
 
 void ssd1306_TestAll() {
     ssd1306_Init();
-    ssd1306_TestFPS();
-    HAL_Delay(3000);
-//    ssd1306_TestBorder();
-    ssd1306_TestFonts();
+//    ssd1306_TestFPS();
+//    HAL_Delay(3000);
+    ssd1306_TestBorder();
+//    ssd1306_TestFonts();
 }
```

This code draws a point that suppose to travel along the border (see ssd1306_TestBorder). It works as expected for SSD1306/SSD1309 OLEDs but for FPGA I see the following behavior: https://www.youtube.com/watch?v=62r7s8MhVHM So there is a bug after all.

Unfortunately I don't understand what causes it and thus how to fix it. Any ideas?

**Comment by afiskon** [comment]
@uXeBoy I managed to fix the issue by shifting X value a little bit:

https://github.com/afiskon/blackice-ii-ssd1306-to-vga/commit/4cd684e70acd8cec8cab30144335dbceb1ad0675

However I still don't have a clear understanding why MAGIC value should be exactly 22. Obviously there is a race condition caused by asynchronous nature of '<=' operator. Which affects not only `raddr` value but also border drawing. It's pretty late in my timezone right now so I'm having difficulties doing the math right now.

Anyway could you please confirm that the new code works for you?

**Comment by uXeBoy** [comment]
Glad that the bug was able to be checked and confirmed, only took two weeks of prodding. :smile:

The math is honestly over my head - my workaround for this bug was to write code that is не слишком изящным! :wink:

**Comment by afiskon** [comment]
@uXeBoy Finally, I solved this mystery!

Here is a full patch https://github.com/afiskon/blackice-ii-ssd1306-to-vga/commit/9af3e65394dd3e3fdfcce726ee22240f5ca1ebfe The description follows.

First of all, turned out that in case of iCEstick everything works as expected, there are no 128th/1th column bugs: https://youtu.be/XizVvOEBizQ

That made me realize two things:
1. I compared SPI configuration for Nucleo and BlackIce II boards and discovered that the SPI configuration of BlackIce II was completely wrong (the the patch).
2. And here is the main gotcha:
```
void ssd1306_Reset(void) {
    HAL_GPIO_WritePin(SSD1306_CS_Port, SSD1306_CS_Pin, GPIO_PIN_SET); // un-select LCD
    /*  
     * Transmit some dummy data with CS = HIGH.
     *
     * This is required to reset `waddr` register on the FPGA side.
     * Since FPGA configuration is done using the same SPI bus and includes
     * sending a few dummy bytes, these bytes cause shifting of `waddr` when
     * the configuration starts.
     */
    uint8_t dummy[1] = {0xAA};
    HAL_SPI_Transmit(&SSD1306_SPI_PORT, dummy, sizeof(dummy), HAL_MAX_DELAY);
```

Here is SystemVerilog part:

```
always_ff @(posedge wclk) // write memory
begin
    if(cs == 0) // chip select
    begin
          mem[waddr] <= din;
          waddr <= waddr + 1;
    end 
    else
        waddr <= 0;
end
```

The problem is that `wclk` is changing only when `cs` is low, thus `waddr` is actually never reset unless we manually send some dummy data with `cs` set to high.

Anyway now I'm pretty sure that everything works as expected in case of both iCEstick and BlackIce II. If you still see some glitches it's worth investigating the library and data it is sending using logical analyzer.
